### PR TITLE
Illumina sample sheet v2

### PR DESCRIPTION
**Pull request by simbig**
- [x] Added automated tests
- [x] Documented for all relevant versions
- [x] Updated the changelog

**Changes**

### Added

- Add support for creating V2 Illumina NovaSeq Sample Sheets

### CODE REVIEW

File: src/IlluminaSampleSheet/V2/BclConvert/DataSection.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        return implode("\n", $bclConvertDataLines) . "\n";
+    }
+
+    /** @param array<string> $samplePropertiesOfFirstSample */
+    private function generateDataHeaderByProperites(array $samplePropertiesOfFirstSample): string
+    {
+        $samplePropertiesOfFirstSample = array_filter($samplePropertiesOfFirstSample, fn ($value) // @phpstan-ignore-next-line Variable property access on a non-object required here
+        => $this->dataRows[0]->$value !== null);
+
+        $samplePropertiesOfFirstSample = array_map(fn ($value) => ucfirst($value), $samplePropertiesOfFirstSample);
```
Comment (spawnia) [inline src/IlluminaSampleSheet/V2/BclConvert/DataSection.php:48]: types missing

File: src/IlluminaSampleSheet/V2/BclConvert/DataSection.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            $bclConvertDataLines[] = implode(',', $dataRow->toArray());
+        }
+
+        return implode("\n", $bclConvertDataLines) . "\n";
+    }
+
+    /** @param array<string> $samplePropertiesOfFirstSample */
+    private function generateDataHeaderByProperites(array $samplePropertiesOfFirstSample): string
+    {
+        $samplePropertiesOfFirstSample = array_filter($samplePropertiesOfFirstSample, fn ($value) // @phpstan-ignore-next-line Variable property access on a non-object required here
```
Comment (spawnia) [inline src/IlluminaSampleSheet/V2/BclConvert/DataSection.php:45]: types missing

File: src/IlluminaSampleSheet/V2/BclConvert/SettingsSection.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+            $bclConvertSettingsLines[] = "TrimUMI,{$booleanToString}";
+        }
+
+        return implode("\n", $bclConvertSettingsLines) . "\n";
+    }
+
+    public function setTrimUMI(bool $trimUMI): SettingsSection
+    {
+        $this->trimUMI = $trimUMI;
```
Comment (spawnia) [inline src/IlluminaSampleSheet/V2/BclConvert/SettingsSection.php:45]: Why is this not also passed through the constructor optionally?

File: src/IlluminaSampleSheet/V2/BclConvert/SettingsSection.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    public function convertSectionToString(): string
+    {
+        $bclConvertSettingsLines = [
+            '[BCLConvert_Settings]',
+            "SoftwareVersion,{$this->softwareVersion}",
+            "FastqCompressionFormat,{$this->fastqCompressionFormat->value}",
+        ];
+
+        if (! is_null($this->trimUMI)) {
+            $booleanToString = $this->trimUMI
```
Comment (spawnia) [inline src/IlluminaSampleSheet/V2/BclConvert/SettingsSection.php:33]: Variable name sounds like it holds a function

File: src/IlluminaSampleSheet/V2/HeaderSection.php [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        $this->instrumentType = $instrumentType;
+
+        return $this;
+    }
+
+    public function setInstrumentPlatform(string $instrumentPlatform): HeaderSection
+    {
+        $this->instrumentPlatform = $instrumentPlatform;
+
+        return $this;
```
Comment (spawnia) [inline src/IlluminaSampleSheet/V2/HeaderSection.php:47]: unnecessary

File: src/IlluminaSampleSheet/V2/HeaderSection.php [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    public function setCustomParam(string $paramName, string $paramValue): void
+    {
+        $this->customParams['Custom_' . $paramName] = $paramValue;
+    }
+
+    public function convertSectionToString(): string
+    {
+        $headerLines = [
+            '[Header]',
+            'FileFormatVersion,' . $this->fileFormatVersion,
```
Comment (spawnia) [inline src/IlluminaSampleSheet/V2/HeaderSection.php:59]:
```suggestion
            "FileFormatVersion,{$this->fileFormatVersion}",
```

File: tests/IlluminaSampleSheet/SampleSheetTest.php [hunk 7]
```diff
@@ -0,0 +1,25 @@
+<?php declare(strict_types=1);
+
+namespace MLL\Utils\Tests\IlluminaSampleSheet;
+
+use MLL\Utils\IlluminaSampleSheet\SampleSheet;
+use MLL\Utils\IlluminaSampleSheet\SectionInterface;
+use PHPUnit\Framework\TestCase;
+
+class SampleSheetTest extends TestCase
```
Comment (spawnia) [inline tests/IlluminaSampleSheet/SampleSheetTest.php:9]:
```suggestion
final class SampleSheetTest extends TestCase
```

File: tests/IlluminaSampleSheet/V2/NovaSeqXCloudSampleSheetTest.php [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+use MLL\Utils\IlluminaSampleSheet\V2\BclConvert\BclSample;
+use MLL\Utils\IlluminaSampleSheet\V2\BclConvert\DataSection;
+use MLL\Utils\IlluminaSampleSheet\V2\BclConvert\SettingsSection;
+use MLL\Utils\IlluminaSampleSheet\V2\Enums\FastQCompressionFormat;
+use MLL\Utils\IlluminaSampleSheet\V2\HeaderSection;
+use MLL\Utils\IlluminaSampleSheet\V2\NovaSeqXSampleSheet;
+use MLL\Utils\IlluminaSampleSheet\V2\ReadsSection;
+use PHPUnit\Framework\TestCase;
+
+class NovaSeqXCloudSampleSheetTest extends TestCase
```
Comment (spawnia) [inline tests/IlluminaSampleSheet/V2/NovaSeqXCloudSampleSheetTest.php:15]:
```suggestion
final class NovaSeqXCloudSampleSheetTest extends TestCase
```

File: tests/IlluminaSampleSheet/V2/NovaSeqXCloudSampleSheetTest.php [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+{
+    public function testNovaSeqXCloudSampleSheetToStringReturnsExpectedResult(): void
+    {
+        $headerSection = new HeaderSection(
+            'Run1',
+        );
```
Comment (spawnia) [inline tests/IlluminaSampleSheet/V2/NovaSeqXCloudSampleSheetTest.php:21]:
```suggestion
        $headerSection = new HeaderSection('Run1');
```

File: tests/IlluminaSampleSheet/V2/NovaSeqXCloudSampleSheetTest.php [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+TrimUMI,0
+
+[BCLConvert_Data]
+Lane,Sample_ID,Index,Index2,OverrideCycles,AdapterRead1,AdapterRead2
+1,Sample1,Index1,Index2,Cycles1,Adapter1,Adapter2
+2,Sample2,Index3,Index4,Cycles2,Adapter3,Adapter4
+3,Sample3,Index5,Index6,Cycles3,Adapter5,Adapter6
+';
+
+        self::assertEquals($expected, $novaSeqXCloudSampleSheet->toString());
```
Comment (spawnia) [inline tests/IlluminaSampleSheet/V2/NovaSeqXCloudSampleSheetTest.php:91]:
```suggestion
        self::assertSame($expected, $novaSeqXCloudSampleSheet->toString());
```

**Review by spawnia — CHANGES_REQUESTED** [review]

File: src/IlluminaSampleSheet/V2/HeaderSection.php (hunk 5 above)
Comment (simbig) [reply src/IlluminaSampleSheet/V2/HeaderSection.php:47]: IMHO this is very opinionated. The fluent interface pattern IHMO is more concise and easy-to-read.
What exactly are the drawbacks that makes it "unnecessary" for you?

File: src/IlluminaSampleSheet/V2/BclConvert/BclSample.php [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+        return $this;
+    }
+
+    public function setBarcodeMismatchesIndex2(string $barcodeMismatchesIndex2): BclSample
+    {
+        $this->barcodeMismatchesIndex2 = $barcodeMismatchesIndex2;
+
+        return $this;
+    }
```
[the remark this answers is not in the document]
Comment (spawnia) [reply src/IlluminaSampleSheet/V2/BclConvert/BclSample.php:75]: All setters.

File: src/IlluminaSampleSheet/V2/ReadsSection.php [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        $this->read2Cycles = $read2Cycles;
+        $this->index1Cycles = $index1Cycles;
+        $this->index2Cycles = $index2Cycles;
+    }
+
+    public function convertSectionToString(): string
+    {
+        $readsLines = ['[Reads]'];
+
+        if (isset($this->read1Cycles)) {
```
Comment (spawnia) [inline src/IlluminaSampleSheet/V2/ReadsSection.php:42]: always set

File: src/IlluminaSampleSheet/V2/ReadsSection.php [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+            $readsLines[] = "Read1Cycles,{$this->read1Cycles}";
+        }
+        if (isset($this->read2Cycles)) {
```
Comment (spawnia) [inline src/IlluminaSampleSheet/V2/ReadsSection.php:45]:
```suggestion
        if ($this->read2Cycles !== null) {
```

File: src/IlluminaSampleSheet/V2/BclConvert/BclSample.php [hunk 14]
```diff
@@ -0,0 +1,112 @@
+<?php declare(strict_types=1);
+
+namespace MLL\Utils\IlluminaSampleSheet\V2\BclConvert;
+
+class BclSample
+{
+    public int $lane;
+
+    public string $sample_ID;
```
[the remark this answers is not in the document]
Comment (spawnia) [reply src/IlluminaSampleSheet/V2/BclConvert/BclSample.php:9]: A comment should explain why we are not using camelCase here.

File: src/IlluminaSampleSheet/V2/BclConvert/BclSample.php [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+
+    public ?string $adapterRead2 = null;
+
+    public ?string $barcodeMismatchesIndex1 = null;
+
+    public ?string $barcodeMismatchesIndex2 = null;
+
+    public function __construct(
+        int $lane,
+        string $sampleID,
```
Comment (spawnia) [inline src/IlluminaSampleSheet/V2/BclConvert/BclSample.php:27]: Should be equal to the property name

**Review by spawnia — APPROVED** [review]

**Pull request merged**